doc/source/release/1.24.0-notes.rst

```diff
@@ -47,9 +47,12 @@ integers with negative values such as ``np.uint8(-1)`` giving
 
 Note that conversion between NumPy integers is unaffected, so that
 ``np.array(-1).astype(np.uint8)`` continues to work and use C integer overflow
-logic.
+logic.  Another approach is using for example ``val % 2**8``.
+When mixing very large integers some care may be necessary since the
+``np.array([large_integers])`` may not return an integer array at this time,
+pass ``np.array(..., dtype=np.int64)`` to guard against this.
```

Comment:
I'm not sure this works as a workaround? I think you get an overflow error for anything larger than `2**63 - 1` and otherwise you'd get `int64` instead of `float64` anyway?

```python
import numpy as np
np.array([np.iinfo(np.uint64).max, 0], dtype=np.int64).astype(np.uint64)
```

```
OverflowError: Python int too large to convert to C long
```

Reply:
Yes, but if that was where you were starting from, the current code either already reads `dtype=np.uint64` or it has no `dtype=` argument.  In both cases you are not affected by the change?!

Reply:
But then does this guard against any of the cases where you'd get `float64`?

Reply:
@adrianeboyd do you have an example of how both of these issues appear at the same time for you?

Reply:
Ah its in the other issue... Well, to be honest, there is no solution then.  You can call `% 2**64` in Python before storing, or something like `(array(..., dtype=obect) % 2*64).astype(np.uint64)`.

That is of course only problematic if you have actual uint64 values and mix them with signed values.  Presumably you can also convert each row because each row with the correct dtype (`uint64` or `int32`) and then cast/concatenate them to the final result?

I can see the issue now, but it still feels pretty niche to me.  And yes, for you this change is _not_ convenient at all.

Reply:
It's definitely on the niche side, but it's still really breaking, since we basically expected casting with the same types in cython and in numpy to do the same thing. Our general usage of numpy is so basic that we've never used upper pins, so this will require a ton of backporting.

I think for the release notes mentioning that `np.array([large_integers])` may return `float64` might still be helpful, but the `dtype=np.int64` suggestion feels more confusing than helpful (to me, anyway). But maybe other users would feel the same way about the `float64` info, I'm clearly not a good judge.